### PR TITLE
[SGD] Add support for native Torch AMP in SGD

### LINKED ISSUE #10919: [RaySGD] mixed precision on PyTorch using torch.cuda.amp
### Describe your feature request
on pytorch >= 1.6.0, torch.cuda.amp could be used for mixed precision training instead of apex (which would reduce a dependency)

PyTorch Lightning allows using either apex or pytorch's native mixed precision. https://pytorch-lightning.readthedocs.io/en/latest/amp.html

[slack](https://ray-distributed.slack.com/archives/CSX7HVB5L/p1598555758016800)

### PR DESCRIPTION
**Pull request by Yard1**
## Why are these changes needed?

Adds support for native automatic mixed precision in Torch RaySGD.

## Related issue number

Will close https://github.com/ray-project/ray/issues/10919

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/util/sgd/torch/training_operator.py [hunk 1]
```diff
@@ -151,15 +151,22 @@ def _set_timers(self, timers):
...
         return models, optimizers
 
     def _configure_ddp(self, models, device_ids, ddp_args):
-        return [
+        models = [
             DistributedDataParallel(model, device_ids=device_ids, **ddp_args)
             for model in models
         ]
+        if self.use_fp16_native:
+            # this is hacky but that's how pytorch lighting does it
```
Comment (amogkam) [inline python/ray/util/sgd/torch/training_operator.py]: This can be removed right?

**Review by amogkam — COMMENTED** [review]
Excellent work! Just left a few questions.

File: python/ray/util/sgd/torch/training_operator.py (hunk 1 above)
Comment (Yard1) [reply python/ray/util/sgd/torch/training_operator.py]: It's been reverted in the second commit, not sure why it's still showing like this

File: python/ray/util/sgd/torch/training_operator.py [hunk 2]
```diff
@@ -133,7 +141,7 @@ def __init__(self,
         self._local_rank = local_rank
         self._config = config
         self._is_distributed = is_distributed
-        self._use_fp16 = use_fp16
+        self._use_fp16 = choose_amp_backend(use_fp16, amp, apex_amp)
```
[the remark this answers is not in the document]
Comment (Yard1) [reply python/ray/util/sgd/torch/training_operator.py:144]: Considering that this attribute could lead to exceptions if set incorrectly, I wanted to play it safe. Plus, if for some unlikely reason someone wants to use just the training operator, the validation logic would be included in it and not dependent on the runner

File: release/nightly_gpu_tests/sgd_gpu/sgd_gpu_test.py [hunk 3]
```diff
@@ -17,7 +17,18 @@
...
-        print(f"The test failed with {e}")
+        print(f"(native fp16) The test failed with {e}")
+        success = False
+
+    try:
+        train_cifar(
+            num_workers=2,
+            use_gpu=True,
+            num_epochs=5,
+            fp16="apex",
```
[the remark this answers is not in the document]
Comment (Yard1) [reply release/nightly_gpu_tests/sgd_gpu/sgd_gpu_test.py:28]: I believe that torch==1.8.1 is used for testing so native amp would be included there

**Comment by Yard1** [comment]
Fixing the release test atm

**Comment by Yard1** [comment]
The sgd release test has been fixed (had some errors in yamls) and set to `num_workers=1` temporarily, otherwise it throws an unrelated error (already reported). If you'd rather have it reverted/disabled, let me know.

**Review by amogkam — APPROVED** [review]
Thanks @Yard1 this looks good! I'm just re-running the flaky rllib build, but will merge once that passes. 

Can you also create an issue to switch `num_workers` back to 2 on the release test so we can track it?

**Comment by richardliaw** [comment]
hmm looks like there's some conflicts in the merge?

**Comment by Yard1** [comment]
@richardliaw should be good now

File: release/sgd_tests/sgd_gpu/sgd_gpu_app_config.yaml [hunk 4]
```diff
@@ -1,4 +1,4 @@
-base_image: "anyscale/ray-ml:pinned-nightly"
+base_image: "anyscale/ray-ml:pinned-nightly-py37-gpu"
```
Comment (amogkam) [inline release/sgd_tests/sgd_gpu/sgd_gpu_app_config.yaml:1]:
```suggestion
base_image: "anyscale/ray-ml:nightly-py37-gpu"
```

**Pull request merged**